src/proxy/http/add_header.rs

```diff
@@ -0,0 +1,229 @@
+use std::{fmt, marker::PhantomData};
+
+use http::header::{AsHeaderName, HeaderValue};
+
+use svc;
+
+type Ret<T> = fn(&T) -> Option<HeaderValue>;
+
+/// Wraps HTTP `Service` `Stack<T>`s so that a given header is removed from a
+/// request or response.
+#[derive(Clone)]
+pub struct Layer<H, T, R> {
+    header: H,
+    retrieve: Ret<T>,
+    _req_or_res: PhantomData<fn(R)>,
+}
+
+/// Wraps an HTTP `Service` so that a given header is added from each request
+/// or response.
+#[derive(Clone)]
+pub struct Stack<H, T, M, R> {
+    header: H,
+    retrieve: Ret<T>,
+    inner: M,
+    _req_or_res: PhantomData<fn(R)>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Service<H, S, R> {
+    header: H,
+    value: HeaderValue,
+    inner: S,
+    _req_or_res: PhantomData<fn(R)>,
+}
+
+// === impl Layer ===
+
+/// Call `request::layer(header)` or `response::layer(header)`.
+fn layer<H, T, R>(header: H, retrieve: Ret<T>) -> Layer<H, T, R>
+where
+    H: AsHeaderName + Clone,
+    R: Clone,
+{
+    Layer {
+        header,
+        retrieve,
+        _req_or_res: PhantomData,
+    }
+}
+
+impl<H, T, M, R> svc::Layer<T, T, M> for Layer<H, T, R>
+where
+    H: AsHeaderName + Clone + fmt::Debug,
+    T: fmt::Debug,
+    M: svc::Stack<T>,
+{
+    type Value = <Stack<H, T, M, R> as svc::Stack<T>>::Value;
+    type Error = <Stack<H, T, M, R> as svc::Stack<T>>::Error;
+    type Stack = Stack<H, T, M, R>;
+
+    fn bind(&self, inner: M) -> Self::Stack {
+        Stack {
+            header: self.header.clone(),
+            retrieve: self.retrieve,
+            inner,
+            _req_or_res: PhantomData,
+        }
+    }
+}
+
+impl<H, T, R> fmt::Debug for Layer<H, T, R>
+where
+    H: fmt::Debug,
+    T: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Layer")
+            .field("header", &self.header)
+            .field("retrieve", &format_args!("{}", "..."))
+            .finish()
+    }
+}
+
+// === impl Stack ===
+
+impl<H, T, M, R> svc::Stack<T> for Stack<H, T, M, R>
+where
+    H: AsHeaderName + Clone + fmt::Debug,
+    T: fmt::Debug,
+    M: svc::Stack<T>,
+{
+    type Value = svc::Either<Service<H, M::Value, R>, M::Value>;
```

Comment:
Why is this necessary? Can't we just always return a wrapped service?

Reply:
I suppose the alternative approach would just be to wrap with a service that has an `Option<HeaderValue>` and just not insert if it's `None`. Might be a little simpler...

-

Reply:
The `Either` style was adapted from the current way we add `client_id` [right now](https://github.com/linkerd/linkerd2-proxy/blob/dc00685639c50c5cf608bb9c0d587b87b9a83282/src/app/inbound.rs#L331). I think changing to an `Option<HeaderValue>` could simplify it, but we are pretty explicit right now when it fails.

Reply:
It's probably fine as it is, I'm just looking for opportunities to simplify.